pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+* **Indexing** - Storing the discovered artifacts in a local in-memory index that enable fast searching.
+* **Searching** ideally using something like the [sublime text's fuzzy search](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), but we definitely need some type of fuzzy search.
+* **Recognition** For several of the above examples (go to definition, find references), we need to be able to recognize what our cursor is hovering over. Is it a module? An atom? a variable? etc.
+
+# Discovery
+Lexical needs a flexible, expandable discovery engine. It needs to be
+able to examine both artifacts on disk like beam files and source
+code, and to index code that's being edited by the user. Right from
+the start, it needs to be able to find the following items:
+
+ * Modules
+ * Public and private module functions
+ * Macros
+ * Anonymous functions
+ * Structs
+ * Records
+ * Types
+ * Variables
+ * Atoms?
+ * Additional items, defined in plugins
+
+While investigating the above items, we will need to take into account
+the difference between usages and definitions. This will be pretty
+clear for most items (though aliasing and importing might muddy the
+waters somewhat), but atoms will be difficult, since there's no
+defintion to speak of, just a usage.
+
+# Caching
+> There are only two hard problems in computer science: Naming things, cache invalidation and off by one errors.
+
+Indexing the entire project, its dependencies, elixir and erlang when
+the language server starts is likely going to be too performance
+intensive to do (to wit, rewriting atoms during namespacing takes 15
+seconds on our moderately sized project), so we will need some sort of
+on-disk representation of the index.
+
+Such a representation will need to take the following into account:
+
+  * Branches can while the language server is running, invalidating large portions of the cache
+  * Project level search and replace can affect large numbers of files at once
+  * The versions of elixir and erlang can change via `asdf` whenever the user wants
+
+Basicallly, the project's assets can change at any time, and it's not
+guaranteed that the language server will be able to detect all of the
+changes, so the language server needs to treat this cache with some
+skepticism and constantly "check its work".
+
+## Cache layout
+A cache entry will have the following format:
+
+```elixir
+
+@type source :: Lexical.uri()
+@entry_type :: :otp | :elixir | :dependency | :project
+
+@type entry %{
+  type: entry_type(),
+  source: source(),
+  entry_version: String.t(),
+  added_at: DateTime.t(),
+  value: term()
+}
+```
+
+### Invalidation
+When lexical starts the first time, it will need to index the
+following entities in the following order:
+
+  1. The elixir release
+  2. All project dependencies
+  3. The project's source code and built artifacts
```

Comment:
Are you referencing something like the compiled beam files of the project itself? Why de we need to index that on top of source files?

One advantage I could see is that macros are already epanded, but I've no idea if that's a valid advantage or how we would map findings for artifacts back to source code.

Reply:
I agree we want to index all 3 "types" of entities, but I'd like to suggest the reverse order so we are able to give suggestions etc sooner to the user (I'm guessing folks will look where a function from project code is defined rather than Elixirs function?). Perhaps we could also speed things up if we're able to ship the index for an Elixir release?

Reply:
Keep in mind that indexing the first two things will happen much less frequently than the third, so I don't think the order will be super relevant to how quickly we serve the index.